Changes in module 'nf-core/falco'
--- modules/nf-core/falco/main.nf
+++ modules/nf-core/falco/main.nf
@@ -33,7 +33,9 @@
         """
     } else {
         """
-        falco $args --threads $task.cpus ${reads}
+        [ ! -f  ${prefix}_1.fastq.gz ] && ln -s ${reads[0]} ${prefix}_1.fastq.gz
+        [ ! -f  ${prefix}_2.fastq.gz ] && ln -s ${reads[1]} ${prefix}_2.fastq.gz
+        falco $args --threads $task.cpus ${prefix}_1.fastq.gz ${prefix}_2.fastq.gz
 
         cat <<-END_VERSIONS > versions.yml
         "${task.process}":

************************************************************
